public/app/features/dashboard-scene/settings/variables/VariableEditorForm.tsx

```diff
@@ -19,28 +19,64 @@ import { VariableNameConstraints } from 'app/features/variables/editor/types';
 import { VariableTypeSelect } from './components/VariableTypeSelect';
 import { EditableVariableType, getVariableEditor, hasVariableOptions, isEditableVariableType } from './utils';
 
+const RESERVED_GLOBAL_VARIABLE_NAME_REGEX = /^(?!__).*$/;
+const WORD_CHARACTERS_REGEX = /^\w+$/;
+
 interface VariableEditorFormProps {
   variable: SceneVariable;
   onTypeChange: (type: EditableVariableType) => void;
   onGoBack: () => void;
   onDelete: (variableName: string) => void;
+  isVariableNameTaken: (name: string, key: string | undefined) => boolean;
 }
-
-export function VariableEditorForm({ variable, onTypeChange, onGoBack, onDelete }: VariableEditorFormProps) {
+export function VariableEditorForm({
+  variable,
+  onTypeChange,
+  onGoBack,
+  onDelete,
+  isVariableNameTaken,
+}: VariableEditorFormProps) {
   const styles = useStyles2(getStyles);
-  const { name, type, label, description, hide } = variable.useState();
+  const [nameError, setNameError] = useState<string | null>(null);
+  const { name, type, label, description, hide, key } = variable.useState();
   const EditorToRender = isEditableVariableType(type) ? getVariableEditor(type) : undefined;
   const [runQueryState, onRunQuery] = useAsyncFn(async () => {
     await lastValueFrom(variable.validateAndUpdate!());
   }, [variable]);
-
   const onVariableTypeChange = (option: SelectableValue<EditableVariableType>) => {
     if (option.value) {
       onTypeChange(option.value);
     }
   };
 
-  const onNameBlur = (e: FormEvent<HTMLInputElement>) => variable.setState({ name: e.currentTarget.value });
+  const onNameChange = useCallback(
+    (e: FormEvent<HTMLInputElement>) => {
+      let errorText = null;
+      if (!RESERVED_GLOBAL_VARIABLE_NAME_REGEX.test(e.currentTarget.value)) {
+        errorText = "Template names cannot begin with '__', that's reserved for Grafana's global variables";
+      }
+
+      if (!WORD_CHARACTERS_REGEX.test(e.currentTarget.value)) {
+        errorText = 'Only word characters are allowed in variable names';
+      }
+
+      if (isVariableNameTaken(e.currentTarget.value, key || '')) {
+        errorText = 'Variable with the same name already exists';
+      }
+
+      if (errorText !== nameError) {
+        setNameError(errorText);
+      }
```

Comment:
```suggestion
      const [isValid, validationError] = onValidateVariableName(name);
      !isValid &&  setNameValidationError(validationError);
```
This logic should be also part of the model